### PR TITLE
multivalue hnsw - uniting hnswlib and hnsw_wrapper

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR replaces the `HirarchicalNSW` class and its `HNSWIndex` wrapper with a single `HNSWIndex` class that derives from `VecSimIndexAbstarct` as the BF algorithm does. This will make the upcoming work on multi-value indexing with the HNSW algorithm much simpler.

These changes do not require any changes in the serialization files. 

**Main objects this PR modified**
1. Uniting the hnswlib and its wrapper functions under the file `hnsw.h`.
2. making the HNSW algorithm a direct part of the VecSim hierarchical tree.
3. Fixing all tests and references to the HNSW algorithm.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 1]
```diff
@@ -50,7 +50,6 @@ class HierarchicalNSW : public VecsimBaseObject {
     double epsilon_;
 
     // Index meta-data (based on the data dimensionality and index parameters)
-    size_t dim;
     size_t data_size_;
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:53]: We need to see which of these fields can be removed... For instance `offsetLevel0_` is always 0.

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 2]
```diff
@@ -311,56 +311,52 @@ void HierarchicalNSW<dist_t>::setIncomingEdgesPtr(idType internal_id, size_t lev
...
+template <typename DataType, typename DistType>
+linklistsizeint *HNSWIndex<DataType, DistType>::get_linklist_at_level(idType internal_id,
+                                                                      size_t level) const {
     return level == 0 ? get_linklist0(internal_id) : get_linklist(internal_id, level);
 }
 
-template <typename dist_t>
-unsigned short int HierarchicalNSW<dist_t>::getListCount(const linklistsizeint *ptr) const {
+template <typename DataType, typename DistType>
+unsigned short int HNSWIndex<DataType, DistType>::getListCount(const linklistsizeint *ptr) const {
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:333]: inline?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 3]
```diff
@@ -166,129 +160,134 @@ class HierarchicalNSW : public VecsimBaseObject {
...
+    int deleteVector(labelType label) override;
+    int addVector(const void *vector_data, labelType label) override;
+    double getDistanceFrom(labelType label, const void *vector_data) const override;
     idType searchBottomLayerEP(const void *query_data, void *timeoutCtx,
                                VecSimQueryResult_Code *rc) const;
-    vecsim_stl::max_priority_queue<dist_t, labelType>
-    searchKnn(const void *query_data, size_t k, void *timeoutCtx, VecSimQueryResult_Code *rc) const;
-    VecSimQueryResult *searchRange(const void *query_data, dist_t radius, void *timeoutCtx,
-                                   VecSimQueryResult_Code *rc) const;
-    bool isLabelExist(labelType label);
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:182]: Why did you remove this?

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 4]
```diff
@@ -1,11 +1,10 @@
 #include "VecSim/algorithms/hnsw/hnsw_factory.h"
-#include "VecSim/algorithms/hnsw/hnsw_wrapper.h"
-#include "VecSim/algorithms/hnsw/hnswlib.h" //linklistsizeint
-#include "VecSim/vec_sim_common.h"          // labelType
+#include "VecSim/algorithms/hnsw/hnsw.h" //linklistsizeint
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw_factory.cpp:2]: I don't like these comments next to the includes...

File: src/VecSim/algorithms/hnsw/serialization.cpp [hunk 5]
```diff
@@ -232,15 +229,7 @@ void HNSWIndexSerializer::loadIndex(const std::string &location) {
     input.close();
 }
 
-void HNSWIndexSerializer::reset(std::shared_ptr<hnswlib::HierarchicalNSW<float>> hnsw_index_) {
-    // Hold the allocator, so it won't be destroyed while VecSimBaseObject is released.
-    auto allocator = hnsw_index->getAllocator();
-    if (hnsw_index_ == nullptr) {
-        hnsw_index.reset();
-        return;
-    }
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/serialization.cpp:241]: Why did you remove this?

File: src/VecSim/algorithms/brute_force/brute_force_single.h [hunk 6]
```diff
@@ -13,9 +13,9 @@ class BruteForceIndex_Single : public BruteForceIndex<DataType, DistType> {
     BruteForceIndex_Single(const BFParams *params, std::shared_ptr<VecSimAllocator> allocator);
     ~BruteForceIndex_Single();
 
-    virtual int addVector(const void *vector_data, size_t label) override;
-    virtual int deleteVector(size_t id) override;
-    virtual double getDistanceFrom(size_t label, const void *vector_data) const override;
+    virtual int addVector(const void *vector_data, labelType label) override;
+    virtual int deleteVector(labelType label) override;
+    virtual double getDistanceFrom(labelType label, const void *vector_data) const override;
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force_single.h:18]: Why virtual?

File: src/VecSim/algorithms/brute_force/brute_force_multi.h [hunk 7]
```diff
@@ -12,11 +12,12 @@ class BruteForceIndex_Multi : public BruteForceIndex<DataType, DistType> {
...
         : BruteForceIndex<DataType, DistType>(params, allocator), labelToIdsLookup(allocator) {}
+
     ~BruteForceIndex_Multi() {}
 
-    virtual int addVector(const void *vector_data, size_t label) override;
-    virtual int deleteVector(size_t id) override;
-    virtual double getDistanceFrom(size_t label, const void *vector_data) const override;
+    virtual int addVector(const void *vector_data, labelType label) override;
+    virtual int deleteVector(labelType labelType) override;
+    virtual double getDistanceFrom(labelType label, const void *vector_data) const override;
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force_multi.h:20]: Why virtual?

**Review by alonre24 — COMMENTED** [review]
Very important PR... added some suggestions for additional refactoring that might be done on the way

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 8]
```diff
@@ -166,129 +160,134 @@ class HierarchicalNSW : public VecsimBaseObject {
     idType getEntryPointId() const;
     labelType getExternalLabel(idType internal_id) const;
     VisitedNodesHandler *getVisitedList() const;
+    virtual VecSimIndexInfo info() const override;
+    virtual VecSimInfoIterator *infoIterator() const override;
+    virtual VecSimBatchIterator *newBatchIterator(const void *queryBlob,
+                                                  VecSimQueryParams *queryParams) override;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:166]: Because they were declared as `virtual` in the wrapper class...

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 9]
```diff
@@ -8,6 +8,8 @@
 #include "VecSim/utils/vec_utils.h"
 #include "VecSim/query_result_struct.h"
 #include "VecSim/vec_sim_common.h"
+#include "VecSim/vec_sim_index.h"
+#include "VecSim/algorithms/hnsw/hnsw_factory.h" //newBatchIterator
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:12]: Ask @meiravgri

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 10]
```diff
@@ -20,23 +22,21 @@
...
-#include <utility> //pair
 
-namespace hnswlib {
 using spaces::dist_func_t;
 
 #define HNSW_INVALID_ID    UINT_MAX
 #define HNSW_INVALID_LEVEL SIZE_MAX
 
-typedef unsigned int linklistsizeint;
+typedef idType linklistsizeint;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:31]: The issue is that the array length is saved in the first place in the array, so the type of the list's length must be of the same type of the ids.

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 11]
```diff
@@ -166,129 +160,134 @@ class HierarchicalNSW : public VecsimBaseObject {
...
 
-template <typename dist_t>
-dist_t HierarchicalNSW<dist_t>::getDistanceByLabelFromPoint(labelType label,
-                                                            const void *data_point) {
-    if (label_lookup_.find(label) == label_lookup_.end()) {
+template <typename DataType, typename DistType>
+double HNSWIndex<DataType, DistType>::getDistanceFrom(labelType label,
+                                                      const void *vector_data) const {
+    auto id = label_lookup_.find(label);
+    if (id == label_lookup_.end()) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:281]: Why?
1. I deleted this function.
2. This way with a single lookup I check for existence and use the id if it exists

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 1 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:53]: I think we should leave this to a proper refactor PR. this value will stay in the base class of hnsw and will be simple to deal with even after the multi-value effort.

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 3 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:182]: Unnecessary. This function was added to check if a label exists in the index from the wrapper class, and now there is no wrapper. We also use the return value of the lookup when we look for a label so there is no point in looking up twice.

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp (hunk 4 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_factory.cpp:2]: @DvirDukhan WDYT? I will align the code when Meirav's PRs will get to the main branch.

File: src/VecSim/algorithms/hnsw/serialization.cpp (hunk 5 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/serialization.cpp:241]: Not needed when using simple pointers. Also, we must use a simple pointer because we interact with the index through the C API.

File: src/VecSim/vec_sim_common.h [hunk 12]
```diff
@@ -191,7 +191,6 @@ typedef int (*timeoutCallbackFunction)(void *ctx);
 typedef enum {
     VecSim_QueryResult_OK = VecSim_OK,
     VecSim_QueryResult_TimedOut,
-    VecSim_QueryResult_Err,
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_common.h:194]: why this one got removed?

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp (hunk 4 above)
Comment (DvirDukhan) [reply src/VecSim/algorithms/hnsw/hnsw_factory.cpp:2]: I agree

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 13]
```diff
@@ -96,199 +92,203 @@ class HierarchicalNSW : public VecsimBaseObject {
...
 
-    HierarchicalNSW() {}                                // default constructor
-    HierarchicalNSW(const HierarchicalNSW &) = default; // default (shallow) copy constructor
-    void setExternalLabel(idType internal_id, labelType label);
-    labelType *getExternalLabelPtr(idType internal_id) const;
-    size_t getRandomLevel(double reverse_size);
-    vecsim_stl::vector<idType> *getIncomingEdgesPtr(idType internal_id, size_t level) const;
-    void setIncomingEdgesPtr(idType internal_id, size_t level, void *set_ptr);
+    HNSWIndex() = delete;                  // default constructor
+    HNSWIndex(const HNSWIndex &) = delete; // default (shallow) copy constructor
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw.h:103]: change the comments please

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 2 above)
Comment (DvirDukhan) [reply src/VecSim/algorithms/hnsw/hnsw.h:334]: inline?

File: src/VecSim/algorithms/hnsw/serialization.cpp (hunk 5 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/serialization.cpp:241]: Two options - using shared pointers instead (requires right use in shared pointers), or using as is "unsafe" and then we should document properly that the index might be freed

File: src/VecSim/vec_sim_common.h (hunk 12 above)
Comment (GuyAv46) [reply src/VecSim/vec_sim_common.h:194]: It was only used for generic return from the try-catch in the wrapper
